divi/src/script/standard.h

```diff
@@ -28,17 +28,24 @@ extern unsigned nMaxDatacarrierBytes;
  */
 static const unsigned int MANDATORY_SCRIPT_VERIFY_FLAGS = SCRIPT_VERIFY_P2SH;
 
+/**
+ * Script flags that are used when verifying the signature on PoS blocks.  They
+ * are consensus relevant as well and cannot be easily changed.
+ */
+static const unsigned int POS_SCRIPT_VERIFY_FLAGS = MANDATORY_SCRIPT_VERIFY_FLAGS |
+                                                    SCRIPT_VERIFY_DERSIG |
+                                                    SCRIPT_VERIFY_STRICTENC |
+                                                    SCRIPT_VERIFY_MINIMALDATA |
+                                                    SCRIPT_VERIFY_NULLDUMMY |
+                                                    SCRIPT_VERIFY_DISCOURAGE_UPGRADABLE_NOPS;
+
 /**
  * Standard script verification flags that standard transactions will comply
  * with. However scripts violating these flags may still be present in valid
  * blocks and we must accept those blocks.
  */
-static const unsigned int STANDARD_SCRIPT_VERIFY_FLAGS = MANDATORY_SCRIPT_VERIFY_FLAGS |
-                                                         SCRIPT_VERIFY_DERSIG |
-                                                         SCRIPT_VERIFY_STRICTENC |
-                                                         SCRIPT_VERIFY_MINIMALDATA |
-                                                         SCRIPT_VERIFY_NULLDUMMY |
-                                                         SCRIPT_VERIFY_DISCOURAGE_UPGRADABLE_NOPS;
+static const unsigned int STANDARD_SCRIPT_VERIFY_FLAGS = POS_SCRIPT_VERIFY_FLAGS |
+                                                         SCRIPT_REQUIRE_COINSTAKE;
```

Comment:
Since the SCRIPT_REQUIRE_COINSTAKE is only needed during the check for the proof-of-stake, why are we altering the STANDARD_SCRIPT_VERIFY_FLAGS? An OP_REQUIRE_COINSTAKE should fail if used anywhere except in a proof-of-stake coinstake transaction.

Reply:
We want the new script rules to be applied for all transactions.  If we were not to do that, then `OP_REQUIRE_COINSTAKE` would precisely *not* fail in non-coinstake transactions, as we would not apply the new rules with this extra check for them.

(The particular part of the code here is just for accepting transactions into the mempool for now, but we definitely also want `OP_REQUIRE_COINSTAKE` inside those to fail, and thus need the flag here.  In the future (after the fork is through), we can think about a simplification of the code where the flag is added to the mandatory flags instead, unless a transaction gets added to the blockchain before the fork that uses the opcode "wrongly".)